File: src/brs/web/api/http/handler/GetAccount.java
package brs.web.api.http.handler;

import static brs.web.api.http.common.Parameters.*;
import static brs.web.api.http.common.ResultFields.*;
import javax.servlet.http.HttpServletRequest;

import brs.web.api.http.common.LegacyDocTag;
import brs.web.api.http.ApiServlet;
import brs.web.api.http.common.JSONData;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import brs.Account;
import brs.Block;
import brs.Blockchain;
import brs.Signum;
import brs.BurstException;
import brs.Constants;
import brs.Generator;
import brs.services.AccountService;
import brs.services.ParameterService;
import brs.util.Convert;
import signumj.crypto.SignumCrypto;

public final class GetAccount extends ApiServlet.JsonRequestHandler {

  private final ParameterService parameterService;
  private final Blockchain blockchain;
  private final AccountService accountService;
  private final Generator generator;

  public GetAccount(ParameterService parameterService, AccountService accountService, Blockchain blockchain, Generator generator) {
    super(new LegacyDocTag[] {LegacyDocTag.ACCOUNTS}, ACCOUNT_PARAMETER, HEIGHT_PARAMETER, GET_COMMITTED_AMOUNT_PARAMETER, ESTIMATE_COMMITMENT_PARAMETER);
    this.parameterService = parameterService;
    this.blockchain = blockchain;
    this.accountService = accountService;
    this.generator = generator;
  }

  @Override
  protected
  JsonElement processRequest(HttpServletRequest req) throws BurstException {

    Account account = parameterService.getAccount(req);

    JsonObject response = JSONData.accountBalance(account);

    int height = parameterService.getHeight(req);
    if(height < 0) {
      height = blockchain.getHeight();
    }

    if(parameterService.getAmountCommitted(req)) {
      long committedAmount = Signum.getBlockchain().getCommittedAmount(account.getId(), height+Constants.COMMITMENT_WAIT, height, null);
      response.addProperty(COMMITTED_NQT_RESPONSE, Convert.toUnsignedLong(committedAmount));
    }

    if(parameterService.getEstimateCommitment(req)) {
      Block block = blockchain.getBlockAtHeight(height);
      long commitment = generator.estimateCommitment(account.getId(), block);
      response.addProperty(COMMITMENT_NQT_RESPONSE, Convert.toUnsignedLong(commitment));
    }

    JSONData.putAccount(response, ACCOUNT_RESPONSE, account.getId());

    if (account.getPublicKey() != null) {
      if(!Convert.checkAllZero(account.getPublicKey())) {
        response.addProperty(ACCOUNT_RESPONSE + "RSExtended", SignumCrypto.getInstance().getAddressFromPublic(account.getPublicKey()).getExtendedAddress());
      }
      response.addProperty(PUBLIC_KEY_RESPONSE, Convert.toHexString(account.getPublicKey()));
    }
    if (account.getName() != null) {
      response.addProperty(NAME_RESPONSE, account.getName());
    }
    if (account.getDescription() != null) {
      response.addProperty(DESCRIPTION_RESPONSE, account.getDescription());
    }

    response.addProperty(IS_AT_RESPONSE, account.isAT());
    response.addProperty(IS_SECURED_RESPONSE, account.getPublicKey() != null);

    if(height == blockchain.getHeight()) {
      // Only if the height is the latest as we don't handle past asset balances.
      // Returning assets here is needed by the classic wallet, so we keep it.
      JsonArray assetBalances = new JsonArray();
      JsonArray unconfirmedAssetBalances = new JsonArray();

      for (Account.AccountAsset accountAsset : accountService.getAssets(account.getId(), 0, -1)) {
        JsonObject assetBalance = new JsonObject();
        assetBalance.addProperty(ASSET_RESPONSE, Convert.toUnsignedLong(accountAsset.getAssetId()));
        assetBalance.addProperty(BALANCE_QNT_RESPONSE, String.valueOf(accountAsset.getQuantityQNT()));
        assetBalances.add(assetBalance);
        JsonObject unconfirmedAssetBalance = new JsonObject();
        unconfirmedAssetBalance.addProperty(ASSET_RESPONSE, Convert.toUnsignedLong(accountAsset.getAssetId()));
        unconfirmedAssetBalance.addProperty(UNCONFIRMED_BALANCE_QNT_RESPONSE, String.valueOf(accountAsset.getUnconfirmedQuantityQNT()));
        unconfirmedAssetBalances.add(unconfirmedAssetBalance);
      }

      if (!assetBalances.isEmpty()) {
        response.add(ASSET_BALANCES_RESPONSE, assetBalances);
      }
      if (!unconfirmedAssetBalances.isEmpty()) {
        response.add(UNCONFIRMED_ASSET_BALANCES_RESPONSE, unconfirmedAssetBalances);
      }
    }

    return response;
  }

}

File: src/main/kotlin/brs/api/http/GetAccount.kt
package brs.api.http

import brs.api.http.common.Parameters.ACCOUNT_PARAMETER
import brs.api.http.common.ResultFields.ACCOUNT_RESPONSE
import brs.api.http.common.ResultFields.ASSET_BALANCES_RESPONSE
import brs.api.http.common.ResultFields.ASSET_RESPONSE
import brs.api.http.common.ResultFields.BALANCE_QUANTITY_RESPONSE
import brs.api.http.common.ResultFields.DESCRIPTION_RESPONSE
import brs.api.http.common.ResultFields.NAME_RESPONSE
import brs.api.http.common.ResultFields.PUBLIC_KEY_RESPONSE
import brs.api.http.common.ResultFields.UNCONFIRMED_ASSET_BALANCES_RESPONSE
import brs.api.http.common.ResultFields.UNCONFIRMED_BALANCE_QUANTITY_RESPONSE
import brs.services.AccountService
import brs.services.ParameterService
import brs.util.convert.toHexString
import brs.util.convert.toUnsignedString
import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import javax.servlet.http.HttpServletRequest

internal class GetAccount internal constructor(
    private val parameterService: ParameterService,
    private val accountService: AccountService
) : APIServlet.JsonRequestHandler(arrayOf(APITag.ACCOUNTS), ACCOUNT_PARAMETER) {

    override fun processRequest(request: HttpServletRequest): JsonElement {

        val account = parameterService.getAccount(request) ?: return JSONResponses.INCORRECT_ACCOUNT

        val response = JSONData.accountBalance(account)
        JSONData.putAccount(response, ACCOUNT_RESPONSE, account.id)

        if (account.publicKey != null) {
            response.addProperty(PUBLIC_KEY_RESPONSE, account.publicKey!!.toHexString())
        }
        if (!account.name.isNullOrEmpty()) {
            response.addProperty(NAME_RESPONSE, account.name)
        }
        if (!account.description.isNullOrEmpty()) {
            response.addProperty(DESCRIPTION_RESPONSE, account.description)
        }

        val assetBalances = JsonArray()
        val unconfirmedAssetBalances = JsonArray()

        for (accountAsset in accountService.getAssets(account.id, 0, -1)) {
            val assetBalance = JsonObject()
            assetBalance.addProperty(ASSET_RESPONSE, accountAsset.assetId.toUnsignedString())
            assetBalance.addProperty(BALANCE_QUANTITY_RESPONSE, accountAsset.quantity.toString())
            assetBalances.add(assetBalance)
            val unconfirmedAssetBalance = JsonObject()
            unconfirmedAssetBalance.addProperty(ASSET_RESPONSE, accountAsset.assetId.toUnsignedString())
            unconfirmedAssetBalance.addProperty(
                UNCONFIRMED_BALANCE_QUANTITY_RESPONSE,
                accountAsset.unconfirmedQuantity.toString()
            )
            unconfirmedAssetBalances.add(unconfirmedAssetBalance)
        }

        if (assetBalances.size() > 0) {
            response.add(ASSET_BALANCES_RESPONSE, assetBalances)
        }
        if (unconfirmedAssetBalances.size() > 0) {
            response.add(UNCONFIRMED_ASSET_BALANCES_RESPONSE, unconfirmedAssetBalances)
        }

        return response
    }
}
